### PR TITLE
Mark Train / Test Split for Study Area

### PR DESCRIPTION
**Pull request by waltaskew**
- For a study area, indicate which 20% of its chunks should be considered the test set and which 80% should be considered the training set.

@tatianawu @prashantkul 
Here's what I was thinking for the test / train split in the metastore. The study area has a list of `{'x_index': x, 'y_index': y}` pairs for the train & test set. So you can look them up, then select the chunks with those indices within the study area you're interested in working with. Does this look all right to y'all?

### CODE REVIEW

**Comment by tatianawu** [comment]
It looks like we're doing the split based on (x, y) indices for each chunk. Does that mean that, across simulations, the same geographic regions are used for training and testing (i.e., each chunk is always used for either training or testing, and we'll never use it for the other)? If so, that could pose an issue.

**Comment by prashantkul** [comment]
I agree. Test data needs to be random enough as training.



*Prashant Kulkarni*

Customer Engineer, Security & Compliance Specialist

Google Cloud

**Comment by waltaskew** [comment]
> It looks like we're doing the split based on (x, y) indices for each chunk. Does that mean that, across simulations, the same geographic regions are used for training and testing (i.e., each chunk is always used for either training or testing, and we'll never use it for the other)? If so, that could pose an issue.

This makes each 1000x1000 chunk either a training chunk or a test chunk.
What should we be doing instead?

> I agree. Test data needs to be random enough as training.

What do you mean by that?

**Comment by waltaskew** [comment]
@prashantkul @tatianawu Can y'all describe what you want for the test train split here? I thought marking 80% of the chunks as 'train' and 20% of the chunks as 'test' is what y'all had requested. Lemme know how you want this to work, and I'll code it up!

**Comment by tatianawu** [comment]
> @prashantkul @tatianawu Can y'all describe what you want for the test train split here? I thought marking 80% of the chunks as 'train' and 20% of the chunks as 'test' is what y'all had requested. Lemme know how you want this to work, and I'll code it up!

We want an 80/20 split on the **examples**, not the chunks. The clarification I was looking for in my original comment was whether these two are synonymous, because I'm concerned they aren't. If we're indexing on (x, y) chunks, does that mean the same geographical chunk is used for either training or testing, across all examples? For example, does (0, 0) represent a well-defined geographical region? And will it always be used for training?

If so, I'm wondering if we can do the split on the labels instead, since I think these more closely resemble the conventional "example"?

**Comment by waltaskew** [comment]
Ah, I see. 

> > @prashantkul @tatianawu Can y'all describe what you want for the test train split here? I thought marking 80% of the chunks as 'train' and 20% of the chunks as 'test' is what y'all had requested. Lemme know how you want this to work, and I'll code it up!
> 
> We want an 80/20 split on the **examples**, not the chunks. The clarification I was looking for in my original comment was whether these two are synonymous, because I'm concerned they aren't. If we're indexing on (x, y) chunks, does that mean the same geographical chunk is used for either training or testing, across all examples? For example, does (0, 0) represent a well-defined geographical region? And will it always be used for training?
> 
> If so, I'm wondering if we can do the split on the labels instead, since I think these more closely resemble the conventional "example"?

Ah, I see. Yes, at present one chunk is a 1000x1000 area and is associated with a single example. The labels describe the same 1000x1000 chunks.

I agree it makes more sense logically to describe the labels rather than the chunks. It'll come out to the same thing, but would be a bit less confusing!

**Comment by tatianawu** [comment]
> > > @prashantkul @tatianawu Can y'all describe what you want for the test train split here? I thought marking 80% of the chunks as 'train' and 20% of the chunks as 'test' is what y'all had requested. Lemme know how you want this to work, and I'll code it up!
> > 
> > 
> > We want an 80/20 split on the **examples**, not the chunks. The clarification I was looking for in my original comment was whether these two are synonymous, because I'm concerned they aren't. If we're indexing on (x, y) chunks, does that mean the same geographical chunk is used for either training or testing, across all examples? For example, does (0, 0) represent a well-defined geographical region? And will it always be used for training?
> > If so, I'm wondering if we can do the split on the labels instead, since I think these more closely resemble the conventional "example"?
> 
> Ah, I see. Yes, at present one chunk is a 1000x1000 area and is associated with a single example. The labels describe the same 1000x1000 chunks.
> 
> I agree it makes more sense logically to describe the labels rather than the chunks. It'll come out to the same thing, but would be a bit less confusing!

Sorry, can you explain how it's the same thing? I think there's some confusion because our data isn't in well-defined examples.

Say we have 2 different simulations (S1 and S2). Looking at a single chunk C1, that is a geographical area with the same geospatial features. However, rainfall looks different between S1 and S2, which means that we have 2 different examples, correct?

Example 1: Uses C1 and S1 rainfall
Example 2: Uses C1 and S2 rainfall

We have 2 different examples that happen to use the same geospatial chunk/features. Am I understanding correctly that the current train/test splitting approach wouldn't make this distinction? It should be possible to have example 1 used for training and example 2 used for testing.

**Comment by waltaskew** [comment]
> > > > @prashantkul @tatianawu Can y'all describe what you want for the test train split here? I thought marking 80% of the chunks as 'train' and 20% of the chunks as 'test' is what y'all had requested. Lemme know how you want this to work, and I'll code it up!
> > > 
> > > 
> > > We want an 80/20 split on the **examples**, not the chunks. The clarification I was looking for in my original comment was whether these two are synonymous, because I'm concerned they aren't. If we're indexing on (x, y) chunks, does that mean the same geographical chunk is used for either training or testing, across all examples? For example, does (0, 0) represent a well-defined geographical region? And will it always be used for training?
> > > If so, I'm wondering if we can do the split on the labels instead, since I think these more closely resemble the conventional "example"?
> > 
> > 
> > Ah, I see. Yes, at present one chunk is a 1000x1000 area and is associated with a single example. The labels describe the same 1000x1000 chunks.
> > I agree it makes more sense logically to describe the labels rather than the chunks. It'll come out to the same thing, but would be a bit less confusing!
> 
> Sorry, can you explain how it's the same thing? I think there's some confusion because our data isn't in well-defined examples.
> 
> Say we have 2 different simulations (S1 and S2). Looking at a single chunk C1, that is a geographical area with the same geospatial features. However, rainfall looks different between S1 and S2, which means that we have 2 different examples, correct?
> 
> Example 1: Uses C1 and S1 rainfall Example 2: Uses C1 and S2 rainfall
> 
> We have 2 different examples that happen to use the same geospatial chunk/features. Am I understanding correctly that the current train/test splitting approach wouldn't make this distinction? It should be possible to have example 1 used for training and example 2 used for testing.

You're right! Yes, this wouldn't capture the distinction between one geography being associated with multiple examples for differing configurations.

**Comment by waltaskew** [comment]
Closing in favor of https://github.com/UrbanSystemsLab/climateiq-cnn/pull/102 which actually does this correctly!
